pkg/networkservice/common/timeout/server.go

```diff
@@ -21,80 +21,145 @@ import (
-	"github.com/networkservicemesh/sdk/pkg/networkservice/core/trace"
 	"github.com/networkservicemesh/sdk/pkg/tools/extend"
+	"github.com/networkservicemesh/sdk/pkg/tools/log"
 )
 
 type timeoutServer struct {
-	onTimeout   *networkservice.NetworkServiceServer
-	connections map[string]*time.Timer
-	executor    serialize.Executor
+	ctx         context.Context
+	connections timerMap
+	executors   executorMap
 }
 
-// NewServer - creates a new NetworkServiceServer chain element that implements timeout of expired connections
-//             - onTimeout - *networkservice.NetworkServiceServer.  Since networkservice.NetworkServiceServer is an interface
-//                        (and thus a pointer) *networkservice.NetworkServiceServer is a double pointer.  Meaning it
-//                        points to a place that points to a place that implements networkservice.NetworkServiceServer
-//                        This is done because when we use timeout.NewServer as part of a chain, we may not *have*
-//                        a pointer to this server used 'onTimeout'.  If we detect we need to heal, onHeal.Request is used to heal.
-//                        If onTimeout is nil, then we simply set onTimeout to this server chain element
-//                        If we are part of a larger chain, we should pass the resulting chain into
-//                        this constructor before we actually have a pointer to it.
-func NewServer(onTimout *networkservice.NetworkServiceServer) networkservice.NetworkServiceServer {
-	rv := &timeoutServer{
-		connections: make(map[string]*time.Timer),
-		onTimeout:   onTimout,
-	}
-	if rv.onTimeout == nil {
-		var actualOnTimeout networkservice.NetworkServiceServer = rv
-		rv.onTimeout = &actualOnTimeout
-	}
-	return rv
+type timer struct {
+	timer  *time.Timer
+	stopCh chan struct{}
 }
 
-func (t *timeoutServer) Request(ctx context.Context, request *networkservice.NetworkServiceRequest) (*networkservice.Connection, error) {
-	ct, err := t.createTimer(ctx, request)
-	if err != nil {
-		return nil, err
+// NewServer - creates a new NetworkServiceServer chain element that implements timeout of expired connections.
+func NewServer(ctx context.Context) networkservice.NetworkServiceServer {
+	return &timeoutServer{
+		ctx: ctx,
 	}
+}
+
+func (t *timeoutServer) Request(ctx context.Context, request *networkservice.NetworkServiceRequest) (conn *networkservice.Connection, err error) {
+	logEntry := log.Entry(ctx).WithField("timeoutServer", "Request")
+
 	connID := request.GetConnection().GetId()
-	t.executor.AsyncExec(func() {
-		if timer, ok := t.connections[connID]; !ok || timer.Stop() {
-			t.connections[connID] = ct
+
+	executor, _ := t.executors.LoadOrStore(connID, &serialize.Executor{})
+	<-executor.AsyncExec(func() {
+		// executor was possibly removed by `t.close()` at this moment, we need to store it back
+		exec, _ := t.executors.LoadOrStore(connID, executor)
+		if exec != executor {
```

Comment:
Why do we need this check?

Reply:
If we have 2 requests executing in parallel we can get such issue:
```
1. -> timeout close  : locking executor
2. -> request-1      : waiting on executor
3. timeout close ->  : unlocking executor, removing it from executors
4. -> request-2      : creating exec, storing into executors, locking exec
5. -request-1->      : locking executor, trying to store it into executors
```
at 5. we get `request-1` locking `executor`, `request-2` locking `exec` and only `exec` stored in `executors`. It means that `request-2` and all subsequent events will be executed in parallel with `request-1`.

Reply:
Yea... LoadOrStore is basically a race... its a well managed race... but still a race.  Simply throwing up your hands and declaring an error is probably a good starting place for handling this situation.  That said, in the future, we *might* think of smarter solutions :)